### ISSUE TITLE
Add Go tests to the CI

### ISSUE THREAD

**Issue by JasonGross** [mit-plv/fiat-crypto#706]
It would be nice to be able to verify that our Go code works.  I'm not sure where to find Go tests, though.  cc @andres-erbsen

**Comment by andres-erbsen** [comment]
I think the most obvious test would be to replace field arithmetic in https://github.com/golang/crypto/blob/master/curve25519/curve25519_generic.go with ours and then run https://github.com/golang/crypto/blob/master/curve25519/curve25519_test.go#L14-L32

I remember it being easier, though, but perhaps I just misremember

**Comment by mdempsky** [comment]
It looks like if you implement the [curve25519 API](https://godoc.org/golang.org/x/crypto/curve25519), it should be easy enough to just run those same tests against it. So I think writing those functions would be the first step.

I'm assuming it would be easy enough to glue the current fiat_25519_* functions together, but could fiat-go generate the glue code itself? E.g., a function like:

    func fiat_25519_scalar_mult(out, scalar, base *[32]byte)

guaranteeing that "out = scalar * base"?

Aside: It looks like the Go backend might need some work still. The "from_byte" and "to_byte" functions are using `[32]uint64` instead of `[32]byte` (or `[32]uint8`).

https://github.com/mit-plv/fiat-crypto/blob/a26d5d416274a05111248165ae7a210194910337/fiat-go/src/curve25519_64.go#L616

https://github.com/mit-plv/fiat-crypto/blob/a26d5d416274a05111248165ae7a210194910337/fiat-go/src/curve25519_64.go#L764

**Comment by mdempsky** [comment]
Go's generic scalarMult code is here: https://github.com/golang/crypto/blob/948cd5f35899cbf089c620b3caeac9b60fa08704/curve25519/curve25519_generic.go

I tried hacking the from_bytes and to_bytes code to operate on bytes instead of uint64, and gluing everything together with the generic scalarMult and feInvert functions along with this bit of adapter glue:

```
type fieldElement = [5]uint64

var (
	feAdd       = fiat_25519_add
	feSub       = fiat_25519_sub
	feMul       = fiat_25519_carry_mul
	feMul121666 = fiat_25519_carry_scmul_121666
	feSquare    = fiat_25519_carry_square

	feFromBytes = fiat_25519_from_bytes
	feToBytes   = fiat_25519_to_bytes
)

func feCopy(dst, src *fieldElement) { *dst = *src }
func feOne(out *fieldElement)       { *out = fieldElement{1} }

func feCSwap(f, g *fieldElement, b int32) {
	b = -b
	for i := range f {
		t := uint64(int64(b)) & (f[i] ^ g[i])
		f[i] ^= t
		g[i] ^= t
	}
}
```

But it seems to be failing all sorts of test vectors. Maybe a different top-level driver loop needs to be used to satisfy fiat-crypto's input bounds?

**Comment by mdempsky** [comment]
Looks like the corresponding BoringSSL code is at https://boringssl.googlesource.com/boringssl/+/master/crypto/curve25519/curve25519.c. Some random spot checking, I don't immediately see any significant differences between the corresponding Go code.

**Comment by mdempsky** [comment]
Using fiat-go's 32-bit code, the test cases work with two changes:

1. Fixing from_bytes and to_bytes to use `[32]byte` instead of `[32]uint32` (mentioned above).
2. Changing from_bytes to mask off the top bit.

Investigating what's wrong with the 64-bit code.

**Comment by mdempsky** [comment]
The 64-bit code is failing because it's mixing up the bits.Mul64 return parameters. The result tuple is `(hi, lo uint64)`, but fiat-go seems to be coded to expect `(lo, hi uint64)`.

Fixing that along with the same changes as for the 32-bit code makes it pass the test vectors.

**Comment by JasonGross** [comment]
> The 64-bit code is failing because it's mixing up the bits.Mul64 return parameters. The result tuple is `(hi, lo uint64)` (https://golang.org/pkg/math/bits/#Mul64), but fiat-go seems to be coded to expect `(lo, hi uint64)`.
> 
> Fixing that along with the same changes as for the 32-bit code makes it pass the test vectors.

Thanks so much for tracking this down!  I'm in the process of regenerating the .go files and pushing a fix now.  The fix is really simple, I just need to swap `hi` and `lo` in
https://github.com/mit-plv/fiat-crypto/blob/a26d5d416274a05111248165ae7a210194910337/src/Stringification/Go.v#L138-L140

> Aside: It looks like the Go backend might need some work still. The "from_byte" and "to_byte" functions are using `[32]uint64` instead of `[32]byte` (or `[32]uint8`).

This is because we choose integer types for bytes (and carries) uniformly across the entire codebase.  We can have `from_bytes` and `to_byes` use `uint8`/`byte`, but then we get errors from the Go compiler about [`bits.Add`](https://golang.org/pkg/math/bits/#Add) needing/returning `uint` rather than `uint8`.  I'm not sure what the right fix is, here.

> I'm assuming it would be easy enough to glue the current fiat_25519_* functions together, but could fiat-go generate the glue code itself? E.g., a function like:
> 
> ```
> func fiat_25519_scalar_mult(out, scalar, base *[32]byte)
> ```
> 
> guaranteeing that "out = scalar * base"?

This is potential future work; the current language we're targeting is only for straightline code, and doesn't really support function calls nor loops.

> 2\. Changing from_bytes to mask off the top bit.

Ah, yes, this is a known incompatibility between our specs and the usage of the functions in BoringSSL / crypto libraries.  I think I recall @andres-erbsen and/or @davidben saying that the proper place to mask off the top bit wasn't really within spec for `from_bytes` (and potentially deserves a different name), but I could be misremembering.

**Comment by JasonGross** [comment]
> > The 64-bit code is failing because it's mixing up the bits.Mul64 return parameters. The result tuple is `(hi, lo uint64)` (https://golang.org/pkg/math/bits/#Mul64), but fiat-go seems to be coded to expect `(lo, hi uint64)`.
> > Fixing that along with the same changes as for the 32-bit code makes it pass the test vectors.
> 
> Thanks so much for tracking this down! I'm in the process of regenerating the .go files and pushing a fix now. The fix is really simple, I just need to swap `hi` and `lo` in

Fixed in e5751ea93

**Comment by mdempsky** [comment]
> but then we get errors from the Go compiler about bits.Add needing/returning uint rather than uint8. I'm not sure what the right fix is, here.

Hm, I'm not seeing this error. Simply removing `--widen-bytes` from here seems to produce code that works, including appropriate `uint8` <-> `uint32`/`uint64` conversions:

https://github.com/mit-plv/fiat-crypto/blob/e5751ea93dd729d55c83eb0896d7f321975bcd2c/Makefile#L207

**Comment by mdempsky** [comment]
> Ah, yes, this is a known incompatibility between our specs and the usage of the functions in BoringSSL / crypto libraries. I think I recall @andres-erbsen and/or @davidben saying that the proper place to mask off the top bit wasn't really within spec for from_bytes (and potentially deserves a different name), but I could be misremembering.

Yeah, I see now that BoringSSL copies the basepoint to mask off the top bit before invoking fiat_25519_from_bytes: https://boringssl.googlesource.com/boringssl/+/master/crypto/curve25519/curve25519.c#164

It should be easy to repeat that within the Go code. I think with your commit (same fix I had in mind) and removing `--widen-bytes`, the generated code should be usable. I'll prepare a commit for upstream and do some benchmarking next.

Is it okay if I send a CL to fix up some stylistic things? E.g., Go code tends to use `//` comments instead of `/*...*/`, and the `/*inline*/` comments don't do anything within Go.

**Comment by JasonGross** [comment]
> Is it okay if I send a CL to fix up some stylistic things? E.g., Go code tends to use `//` comments instead of `/*...*/`, and the `/*inline*/` comments don't do anything within Go.

Yes, please do.  And let me know if there's anything I can help with here.

**Comment by JasonGross** [comment]
> Hm, I'm not seeing this error. Simply removing `--widen-bytes` from here seems to produce code that works, including appropriate `uint8` <-> `uint32`/`uint64` conversions:

https://github.com/mit-plv/fiat-crypto/pull/843/checks?check_run_id=861762722#step:5:20 in https://github.com/mit-plv/fiat-crypto/pull/843
```
fiat-go/src/p256_64.go:322:25: cannot use x183 (type uint64) as type uint8 in argument to fiat_p256_cmovznz_u64
fiat-go/src/p256_64.go:324:25: cannot use x183 (type uint64) as type uint8 in argument to fiat_p256_cmovznz_u64
fiat-go/src/p256_64.go:326:25: cannot use x183 (type uint64) as type uint8 in argument to fiat_p256_cmovznz_u64
fiat-go/src/p256_64.go:328:25: cannot use x183 (type uint64) as type uint8 in argument to fiat_p256_cmovznz_u64
fiat-go/src/p256_64.go:613:25: cannot use x183 (type uint64) as type uint8 in argument to fiat_p256_cmovznz_u64
fiat-go/src/p256_64.go:615:25: cannot use x183 (type uint64) as type uint8 in argument to fiat_p256_cmovznz_u64
fiat-go/src/p256_64.go:617:25: cannot use x183 (type uint64) as type uint8 in argument to fiat_p256_cmovznz_u64
fiat-go/src/p256_64.go:619:25: cannot use x183 (type uint64) as type uint8 in argument to fiat_p256_cmovznz_u64
fiat-go/src/p256_64.go:670:25: cannot use x18 (type uint64) as type uint8 in argument to fiat_p256_cmovznz_u64
fiat-go/src/p256_64.go:672:25: cannot use x18 (type uint64) as type uint8 in argument to fiat_p256_cmovznz_u64
```

I'll dig into this a bit more...

**Comment by mdempsky** [comment]
Ah, I was only testing curve25519_{32,64}.go. Yeah, I see those errors with some of the other files.

**Comment by JasonGross** [comment]
Okay, I think I've tracked down the issue.  The problem is that I'm synthesizing `cmovznz` and telling it that the `bytes` size is allowed at https://github.com/mit-plv/fiat-crypto/blob/e5751ea93dd729d55c83eb0896d7f321975bcd2c/src/PushButtonSynthesis/Primitives.v#L842-L850, but when I synthesize, e.g., `opp` at https://github.com/mit-plv/fiat-crypto/blob/e5751ea93dd729d55c83eb0896d7f321975bcd2c/src/PushButtonSynthesis/WordByWordMontgomery.v#L347-L355 I tell it that the `bytes` size is not allowed.  This hasn't caused issues before, because I don't think I have any other examples where carries should be widened to something wider than bytes.  I'm working on fixing this descrepency now.

**Comment by mdempsky** [comment]
Out of curiosity, why wasn't that caught by the Coq correctness proofs?

**Comment by JasonGross** [comment]
This part of the pipeline doesn't yet have a correctness proof about it, unfortunately.  The proofs ensure that the integer types are such that truncating down to that integer type doesn't result in incorrectly computed values.  However, the final pass, where we pretty-print the Go code and introduce integer types rather than ranges, doesn't have any proofs about it.

**Comment by JasonGross** [comment]
Argh, this is a pain.  Some primitives (e.g., add-with-carry, sub-with-borrow) are used both in functions that permit bytes, and in functions that do not permit bytes sizes.  I don't see a good way to synthesize this code, because there doesn't seem to be a uniform rule about how to handle integer sizes.  Do you have ideas?

Alternatively, is there a way to replace `bits.Add64` and `bits.Add32` with some construction that turns the carry into a uint8?  (And then we don't have to widen carries at all.)  Like, `(uint64, uint8)(bits.Add64(...))` or something like that?  Alternatively, I could go back to including a thin wrapper around these functions, like
```
//
// The function fiat_25519_Add64 is a thin wrapper around bits.Add64 that uses uint8 rather than uint64
//
func fiat_25519_Add64(x uint64, y uint64, carry uint8) {
  var sum uint64
  var carryOut uint64
  sum, carryOut = bits.Add64(x, y, carry)
  return sum, uint8(carryOut)
}
```

**Comment by mdempsky** [comment]
> I don't see a good way to synthesize this code, because there doesn't seem to be a uniform rule about how to handle integer sizes. Do you have ideas?

Do you mean no uniformity within Go or within fiat-crypto?

Go's rules about integer sizes are fairly simple. Integer literals are "untyped" exact constants, until given a specific type (e.g., explicitly converting to some type, assigning to a typed-variable, or including in an arithmetic expression with another typed value). And there's no implicit conversion between typed values; e.g., a uint8 variable has to be explicitly converted to uint64.

> Alternatively, I could go back to including a thin wrapper around these functions, like

That sounds fine to me if it's simplest. The Go compiler should be able to inline and optimize away the extra conversions. (And if not, it's probably a sign we should tweak the inlining heuristics.)

**Comment by mdempsky** [comment]
I uploaded a work-in-progress CL at https://go-review.googlesource.com/c/crypto/+/242177 to import the curve25519 code into Go's standard library. It gives a significant performance improvement over the status quo generic implementation: 38% faster on 386, and 61% faster on amd64. (Caveat: Go has an assembly implementation for amd64 that's still faster and will normally be used instead.)

**Comment by JasonGross** [comment]
> Do you mean no uniformity within Go or within fiat-crypto?

I mean uniformity in the function type signatures in the Go standard library.  For example, four possible rules are "each variable should be typed as the smallest integer size that can hold all valid values", "each variable should be typed as the smallest integer size out of 8, 32, and 64 bits that can hold all valid values (so, no 16-bit variables)", "each variable should be typed as the smallest integer size out of 32 and 64 bits that can hold all valid values", and "all variables are 64 bits".  The Go standard library picks none of these conventions.  In particular, it does not pick "tightest from 8, 32, 64 bits" because bits.Add32 and bits.Add64 violate this convention (the carry variable is uint, not uint8), and it does not pick "tightest from 32, 64 bits" because to_bytes and from_bytes use uint8, not uint.

> That sounds fine to me if it's simplest. 

Great, I'll plan to implement this within the next day or two

**Comment by mdempsky** [comment]
> The Go standard library picks none of these conventions. 

I see. The math/bits package follows the convention that the `FooNN` function accepts and returns only `uintNN` (eg, `Add32` uses only `uint32`, and `Add64` uses only `uint64`).

The undecorated `Foo` functions use just `uint` but I don't think you're using any of those variants anyway.

**Comment by JasonGross** [comment]
@mdempsky The code in https://github.com/mit-plv/fiat-crypto/pull/843 should work now.  It's a bit uglier, since we define a typedef for the carry type (see #846 for a bit more context).  If you want me to instead inline `uint8` instead of defining `fiat_25519_uint1`, I can do that.

**Comment by mdempsky** [comment]
@JasonGross Defining a `uint1` type is fine by me. Shouldn't affect code generation at all.

(You can also generate type declarations unconditionally if you want. Go only disallows importing unused packages and declaring unused function-scoped variables.)

**Comment by mdempsky** [comment]
Last night I was playing with making use of fiat-crypto's NIST P-224 code. The fiat-crypto 32-bit code seems marginally slower than Go's existing [constant-time P-224 implementation](https://golang.org/src/crypto/elliptic/p224.go) on 386 (caveat: I realized later I wasn't controlling for CPU throttling, and the difference is small enough that this could be responsible), but the 64-bit code is about twice as fast on amd64.

A couple notes on the generated NIST curve code:

1. The `fiat_p224_from_bytes` comment says it deserializes "a field element in the Montgomery domain", but this seems to be incorrect. You actually have to use `fiat_p224_from_bytes` to convert `*[28]byte` to `*[7]uint32`, and then use `fiat_p224_to_montgomery` to convert to Montgomery domain.

   I don't think Go needs these two operations separately, so they could be fused; but I also don't imagine any performance benefit from this. So maybe just fixing the comment would be sufficient? (Caveat: my understanding of EC crypto is limited to basic group theory; I don't really understand "Jacobi" or "Montgomery", etc., so maybe the current comment is actually correct.)

2. The 64-bit `fiat_p224_from_bytes` and `fiat_p224_to_bytes` functions take `*[32]byte`, but they should take `*[28]byte`.

3. The point [addition](https://hyperelliptic.org/EFD/g1p/data/shortw/jacobian-3/addition/add-2007-bl)/[doubling](https://hyperelliptic.org/EFD/g1p/data/shortw/jacobian-3/doubling/dbl-2001-b) formulas Go uses involve some small fixed scalar multiplications: 2x, 3x, 4x, and 8x (often in contexts like x-8y). The existing Go code handles these specially with bit shifting. With fiat-go, I implemented them as sequences of `fiat_p224_add` calls.

   Assuming it's not too much more work to generate small linear-arithmetic combinations like this, it might be slightly faster if fiat-go were to generate code for these, like it the 121666 multiplication for curve25519?

**Comment by JasonGross** [comment]
> 1\. The `fiat_p224_from_bytes` comment says it deserializes "a field element in the Montgomery domain", but this seems to be incorrect. You actually have to use `fiat_p224_from_bytes` to convert `*[28]byte` to `*[7]uint32`, and then use `fiat_p224_to_montgomery` to convert to Montgomery domain.
>    I don't think Go needs these two operations separately, so they could be fused; but I also don't imagine any performance benefit from this. So maybe just fixing the comment would be sufficient?

Ah, that seems right.  Perhaps just "[...] encodes an integer as a field element NOT in the Montgomery domain"?

> 2\. The 64-bit `fiat_p224_from_bytes` and `fiat_p224_to_bytes` functions take `*[32]byte`, but they should take `*[28]byte`.

This is a result of the number of bytes-limbs being computed from the number of limbs; 4 64-bit limbs is the same as 32 8-bit limbs.  I will try adjusting https://github.com/mit-plv/fiat-crypto/blob/611908f4d8796d632aaca78582f45dc3ad4d2022/src/Arithmetic/Freeze.v#L183-L185 to fix this.
However, I was under the impression that Montgomery reduction was only efficient when using saturated limbs, and that unsaturated Solinas should be used instead for unsaturated primes (unfortunately, unsaturated Solinas can't yet be used for p224 because of #554).  I'm wondering if @andres-erbsen has thoughts on this state of affairs (and whether or not we should allow Montgomery to better handle unsaturated cases.)  Pull request coming soon.

> 3\. Assuming it's not too much more work to generate small linear-arithmetic combinations like this, it might be slightly faster if fiat-go were to generate code for these, like it the 121666 multiplication for curve25519?

Done in https://github.com/mit-plv/fiat-crypto/pull/847

**Comment by mdempsky** [comment]
> However, I was under the impression that Montgomery reduction was only efficient when using saturated limbs, and that unsaturated Solinas should be used instead for unsaturated primes (unfortunately, unsaturated Solinas can't yet be used for p224 because of #554).

I'm certainly no expert here. I was just trying to plug together the fiat-go primitives I had available.

In particular, my observation was fiat-go's add/sub/mul/square routines's postconditions were all written in terms of montgomery evaluation, and I wasn't able to get the test vectors to pass except by using `from_bytes`+`to_montgomery` at the beginning, and `from_montgomery`+`to_bytes` at the end.
